Yeksi Naa Reports | Satisfaction Rate after delivery report

dmydlo opened the pull request:
Hi @czue,
This is an 8th Yeksi Naa Report, Satisfaction Rate after delivery report.

Commit:
Yeksi Naa Reports | Satisfaction Rate after delivery report

Commit:
Yeksi Naa Reports | Test travis

Commit:
Yeksi Naa Reports | Add sorting in dashboard 3 test case

czue commented on custom/yeksi_naa_reports/sqldata.py:
```diff
@@ -809,6 +809,87 @@ def headers(self):
+            total_row.append(self.percent_fn(
+                sum(
+                    products[product_id][i]['numerator'] for product_id in products if
+                    products[product_id][i]['numerator'] and products[product_id][i]['denominator']
```
maybe not important, but i'm curious why a numerator is required here? 

wouldn't 0/100 be a relevant data point to include in a total percentage?

dmydlo replied:
Right, it should only check denominator existence to include only relevant data.

czue commented on custom/yeksi_naa_reports/sqldata.py:
```diff
@@ -809,6 +809,87 @@ def headers(self):
+        data = {}
+        product_names = {}
+        for row in rows:
+            if self.months[0] <= row['real_date_repeat'] < self.months[-1] + relativedelta(months=1):
```
nit: it would be much easier to parse this if it was extracted into a function. e.g. `if self._date_is_in_range(row['real_date_repeat'])` (or whatever an appropriate name would be)

czue commented on custom/yeksi_naa_reports/sqldata.py:
```diff
@@ -809,6 +809,87 @@ def headers(self):
+                        data[row['product_id']].append({
+                            'numerator': 0,
+                            'denominator': 0,
+                        })
```
this could probably be simplified / eliminated if you used a couple nested `defaultdict`s

czue commented on custom/yeksi_naa_reports/sqldata.py:
```diff
@@ -809,6 +809,87 @@ def headers(self):
+        rows = self.get_data()
+        data = {}
+        product_names = {}
+        for row in rows:
```
this entire loop could maybe use a comment explaining what it's doing. or maybe better be extracted into a function with a name explaining what it's doing.

czue commented on custom/yeksi_naa_reports/sqldata.py:
```diff
@@ -809,6 +809,87 @@ def headers(self):
+                        break
+
+        new_rows = []
+        for product_id in data:
```
(same comment as other loop)

czue commented on custom/yeksi_naa_reports/sqldata.py:
```diff
@@ -809,6 +809,87 @@ def headers(self):
+        for product_id in data:
+            row = [product_names[product_id]]
+            for i in range(len(self.months)):
+                if data[product_id][i]['numerator'] and data[product_id][i]['denominator']:
```
same question as above - is it not ever valid to have a numerator of `0` but a real denominator?

czue commented on custom/yeksi_naa_reports/tests/reports/test_dashboard_3.py:
```diff
@@ -8,6 +8,61 @@
 
 class TestDashboard3(YeksiTestCase):
 
+    def test_satisfaction_rate_after_delivery_data_report(self):
```
:+1:

czue reviewed (commented):
most/all of my comments are not blockers, but want to hear your response on the exclusion of numerators of 0 before merging.

Commit:
Yeksi Naa Reports | Name loops in Satisfaction rate after delivery report, rename rows to records, create helper functions for operations with dates

dmydlo commented:
After rebasing with the master branch I have named loops in Satisfaction Rate after a delivery report as well as I've added use of defaultdict's.
On top of that, I have renamed rows to records across all sqldata as these come from database and we are returning table rows and added usage of created helper functions for operating on dates also across the whole document.

If you have any questions I will be happy to answer.
P.S. Thanks for great bits of advice.

czue commented on custom/yeksi_naa_reports/sqldata.py:
```diff
@@ -38,6 +39,14 @@ def months(self):
+    def get_index_of_month_in_selected_data_range(self, date):
+        for index in range(len(self.months)):
+            if date < self.months[index] + relativedelta(months=1):
+                return index
```
is it ever possible that this condition would never be hit and this function would return `None`? It seems like a lot of code below would break if that ever happened, so could be good to be explicit (e.g. by throwing an exception outside the bottom of the loop if it reaches there).

dmydlo replied:
It will never hit `None` as before each use we check `self.date_in_selected_date_range`, which checks if it would hit `None`

czue replied:
got it, thanks

czue reviewed (approved):
thanks for the cleanup! I found the revised version much easier to read/reason with - hopefully you did too. 

I had one last thought/question but I'll go ahead and approve/merge this and let you follow up as/if you see fit.
